File: api/src/services/recipe/recipe-service.ts
import { IRecipeRecord } from '../../repositories/recipe/recipe-repository.interface.js';
import { RecipeRepository } from '../../repositories/recipe/recipe-repository.js';
import { Connection } from '../../utils/database.js';
import { DBService } from '../db-service.js';

/**
 * Recipe Service class.
 *
 * @class RecipeService
 * @property {RecipeRepository} repository - Recipe repository dependency
 */
export class RecipeService implements DBService {
  repository: RecipeRepository;

  constructor(connection: Connection) {
    this.repository = new RecipeRepository(connection);
  }

  /**
   * Get recipe.
   *
   * @async
   * @param {number} cookbookId
   * @returns {Promise<IRecipeRecord>}
   */
  async getRecipe(cookbookId: number): Promise<IRecipeRecord> {
    // TODO: This should also get the ingredients
    return this.repository.getRecipeById(cookbookId);
  }

  /**
   * Get cookbook recipes.
   *
   * @async
   * @param {number} cookbookId
   * @returns {Promise<IRecipeRecord[]>}
   */
  async getCookbookRecipes(cookbookId: number): Promise<IRecipeRecord[]> {
    return this.repository.getRecipesByCookbookId(cookbookId);
  }
}

File: backend/api/service/recipeService.go
package service

import (
	"database/sql"
	"recipevault/api/repository"
	"recipevault/models"
)

type RecipeService struct {
	repository        *repository.RecipeRepository
	ingredientService *IngredientService
}

// Construct new instance of RecipeService
func NewRecipeService(r *repository.RecipeRepository, i *IngredientService) *RecipeService {
	return &RecipeService{
		repository:        r,
		ingredientService: i,
	}
}

// Initialize RecipeService with dependencies
func InitRecipeService(db *sql.DB) *RecipeService {
	r := repository.NewRecipeRepository(db)
	i := InitIngredientService(db)
	return NewRecipeService(r, i)
}

// Get recipe
func (s *RecipeService) GetRecipe(recipeID int) (*models.RecipeIngredients, error) {
	r, err := s.repository.GetRecipeByID(recipeID)

	if err != nil {
		return nil, err
	}

	i, err := s.ingredientService.GetRecipeIngredients(recipeID)

	if err != nil {
		return nil, err
	}

	recipe := &models.RecipeIngredients{
		Recipe:      r,
		Ingredients: i,
	}

	return recipe, nil
}

// Get cookbook recipes
func (s *RecipeService) GetCookbookRecipes(cookbookID int) ([]models.Recipe, error) {
	return s.repository.GetRecipesByCookbookID(cookbookID)
}
